### ISSUE TITLE
cross-origin requests fail

### ISSUE THREAD

**Issue by dan-reznik** [opencpu/opencpu#312]
A web application makes a cross-origin HTTP request when it requests a resource that has a different origin (domain, protocol, and port) than its own origin, e.g., some frontend JavaScript code served from https://codepen.io/dan-reznik-the-looper/full/yqapyw/ uses fetch() to make a request for "https://dan-reznik.ocpu.io/AzorPkg2/json/tree_ale_history_grouped.json".

fetch() follows a [same-origin](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy) policy. This means that browsers restrict cross-origin HTTP requests from within scripts. 

To circumvent this the _server_ the server should add the header 'Access-Control-Allow-Origin: *' to the http response. I understand openCPU currently echoes the origin, and I do think this is the proper default behavior, however, when requesting a json (images don't generate this problem as they are exceptions) the default header behavior will produce the cross-origin error on modern browsers.

There's a free service called "cors-anywhere" which allows you to issue a POST request to an endpoint passing the desired requested payload (in my case a json) as a parameter, as well as the _desired_ value to be inserted in the response header. See "cors-opt" variable below. So this acts as an intermediary proxy between the resource's location (at opencpu in this case) and the caller.

```{js}
const url = "https://dan-reznik.ocpu.io/AzorPkg2/json/tree_ale_history_grouped.json";
const url_base = "https://dan-reznik.ocpu.io";
const cors_url = 'https://cors-anywhere.herokuapp.com/';
const cors_opts = {
  method: 'get',
  headers: {'Origin': url_base}
};
// to avoid cross-origin
// fetch(cors_url+url, cors_opts)
fetch(url)
  .then(function(response) {
    return response.json();
  })
  .then(function(myJson) {
// do something
    return(myJson);
  });
```
the request here would be for the openCPU service to allow us to pass one extra paremeter (call it 'Origin') on a POST and a GET whereby we can instruct openCPU to have the non-default behavior of letting the requester define what value should be on the response's 'Access-Control-Allow-Origin' header field, e.g., by fixing it to a fixed url and therefore averting the cross-origin error.

**Comment by jeroen** [comment]
You can already do this, by setting the `origin` request header.

**Comment by dan-reznik** [comment]
Jeroen I will try that. But I full expect the origin portion of the request header to be slapped onto the request by the browser (which interprets js) after request is issued, i.e., it is not under the caller's control. I will still try it and let you know. Again, if this simple workaround were possible none of the of pages which recommend using the cors-anywhere service would exist.

**Comment by dan-reznik** [comment]
Origin is a 'forbidden' header, meaning that you cannot change it programatically. You would need to configure the web server to allow CORS requests. https://stackoverflow.com/questions/46808011/how-to-set-the-origin-request-header

**Comment by dan-reznik** [comment]
I am staging a few data files (as .json's) which hang off of my /inst directory on my R package, for convenience, as they are produced by R code. It is true I could host them elsewhere but they are intrinsically connected to the logic of my application. So this request is one of convenience.

**Comment by jeroen** [comment]
I am still not convinced adding an extra parameter is the solution. OCPU doesn't do anything special, any web service that supports CORS does the same thing. I don't see why opencpu would need to support an extra hack for this.

Can you show me the application that doesn't work so I can understand why the browser is sending the wrong `origin` header? I am not sure your diagnosis of the problem is entirly correct.

**Comment by dan-reznik** [comment]
i will try to reproduce that error and send you a screenshot. What I can tell you is the problem went away immediately with cors-anywhere when i was able to set the return header's Origin field (as a parameter to cors-anywhere).

**Comment by jeroen** [comment]
Yes but that is a workaround, it does not explain why the problem appears in the first place. The browser should always send the correct origin header. So if there is a bug somewhere in opencpu that messes this up I would like to know about it.

**Comment by dan-reznik** [comment]
The error arose because I was running two apps on separate tabs of my
browser. The first one at surge.sh and the 2nd one at codepen.io.
Significantly, both were requesting the same resource (.json hosted at
opencpu) and that led to the error. As soon as I placed the cors-anywhere
service as a proxy the error went away! It's simple.

Dan

**Comment by dan-reznik** [comment]
will do. i am flying now using my phone. whenever i am on the ground i will
do this

**Comment by dan-reznik** [comment]
hi jeroen. here's some experimentation in response to "
Could you check in devtools in the response headers about the timestamp of
the response and the x-ocpu-cache response header?"

a first test: requesting the .json directly from the browser (i was
suprised to see access-control-allow-origin coming back as "*", which
weakens my diagnosis as the true source of the error. There could indeed be
some caching issues. I am trying to figure out a way to make a JS app
running on another server, fetching the resource from ocpu to report the
response headers, when I get that I will send you another email


   1. Request URL:
      https://dan-reznik.ocpu.io/AzorPkg2/json/tree_ale.json
      2. Request Method:
      GET
      3. Status Code:
      200
      4. Remote Address:
      45.79.171.71:443
      5. Referrer Policy:
      no-referrer-when-downgrade
      1. Response Headers
      1. access-control-allow-credentials:
      true
      2. access-control-allow-headers:
      Origin, Content-Type, Accept, Accept-Encoding, Cache-Control,
      Authorization
      3. access-control-allow-origin:
      *
      4. access-control-expose-headers:
      Location, X-ocpu-session, Content-Type, Cache-Control
      5. cache-control:
      max-age=900, public
      6. content-encoding:
      gzip
      7. content-length:
      16736
      8. content-type:
      application/json
      9. date:
      Sun, 29 Jul 2018 14:25:18 GMT
      10. server:
      nginx/1.12.0
      11. status:
      200
      12. vary:
      Accept-Encoding
      13. x-ocpu-cache:
      BYPASS
      14. x-ocpu-locale:
      en_US.UTF-8
      15. x-ocpu-r:
      R version 3.4.4 (2018-03-15)
      16. x-ocpu-server:
      rApache
      17. x-ocpu-time:
      2018-07-29 14:25:18 UTC
      18. x-ocpu-version:
      2.0.8
      2. Request Headers
      1. :authority:
      dan-reznik.ocpu.io
      2. :method:
      GET
      3. :path:
      /AzorPkg2/json/tree_ale.json
      4. :scheme:
      https
      5. accept:

      text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
      6. accept-encoding:
      gzip, deflate, br
      7. accept-language:
      en-US,en;q=0.9,pt-BR;q=0.8,pt;q=0.7
      8. cache-control:
      no-cache
      9. dnt:
      1
      10. pragma:
      no-cache
      11. upgrade-insecure-requests:
      1
      12. user-agent:
      Mozilla/5.0 (X11; CrOS x86_64 10575.58.0) AppleWebKit/537.36 (KHTML,
      like Gecko) Chrome/67.0.3396.99 Safari/537.36


2) requesting it from a small stackblitz app (just to send the originating
server)

**Comment by dan-reznik** [comment]
a) when fetch resource directly from browser url field, namely

https://dan-reznik.ocpu.io/AzorPkg2/json/tree_ale.json

access-control-allow-origin comes back as "*"

b) following that, running a js script hosted in some server
(stackblitz.com), the above success is successfully fetched, and
access-control-allow-origin comes back as
"some-preamble.stackblitz.com"

c) finally, when the exact same script in (b) is run from within
codepen.io, the fetch fails (cross-origin error), and
access-control-allow-origin response header comes back as
"https://s.codepen.io".

do you see any light?

**Comment by jeroen** [comment]
At what URL is the failing page being hosted then? If it sends `Origin: https://s.codepen.io` but you say it is actually hosted at `https://codepen.io` ?

One thing I noticed is that the webserver at `https://codepen.io` is setting an invalid value for the `X-Frame-Options` response header. This could also be the reason that the browser blocks cross domain requests.

**Comment by dan-reznik** [comment]
it fails at the 2nd step of the following sequence:

(a) a first (and successful) js fetch() is done from a first server (in my
case stackblitz.com)

(b) a 2nd js fetch() for same resource is done from a 2nd server (codepen.io).
at this point the browser produces the cross-origin error.

**Comment by dan-reznik** [comment]
it fails at the 2nd step of the following sequence:

(a) a first (and successful) js fetch() is done from a first server (in my case stackblitz.com)

(b) a 2nd js fetch() for same resource is done from a 2nd server (codepen.io). **at this point the browser produces the cross-origin error.**
